pymodbus/transport/transport_serial.py

```diff
@@ -51,7 +51,7 @@ def close(self, exc: Exception | None = None) -> None:
         else:
             self.async_loop.remove_reader(self.sync_serial.fileno())
         self.sync_serial.close()
-        self.sync_serial = None
+        del self.sync_serial
```

Comment:
This works only in theory as far as I can see, for a couple of reasons:
- the variables set in __init__ in sync_serial is also removed
- it prohibits reusing the object, because sync_serial no longer exist
- It makes debugging quite difficult because you can no longer see the (faulty) sync_serial.

Reply:
> the variables set in init in sync_serial is also removed

What variables do you expect to persist after `close()` is called, and why would they need to persist?

> it prohibits reusing the object, because sync_serial no longer exist

The existing code doesn't allow re-use, since you can't any method after it's been set to `None`
`AttributeError: 'NoneType' object has no attribute 'foo'`
Any re-use will have to recreate it anyways.

> It makes debugging quite difficult because you can no longer see the (faulty) sync_serial.

I'd rather deal with an explicitly deleted object rather than one that may be garbage collected at any time!

Reply:
after calling close() you can call connect().

I politely disagree, deleting objects is not really standard python.

Reply:
> after calling close() you can call connect().

```python
   self.async_loop.remove_reader(self.sync_serial.fileno())
        self.sync_serial.close()
        self.sync_serial = None
>       self.sync_serial.connect()  # not in dev, for demonstration only
E       AttributeError: 'NoneType' object has no attribute 'connect'

pymodbus/transport/transport_serial.py:55: AttributeError
======================================= short test summary info =======================================
FAILED test/sub_transport/test_basic.py::TestBasicSerial::test_external_methods - AttributeError: 'NoneType' object has no attribute 'connect'
FAILED test/sub_transport/test_comm.py::TestCommModbusProtocol::test_connected[CommType.SERIAL-socket://localhost:5020] - AttributeError: 'NoneType' object has no attribute 'connect'
FAILED test/sub_transport/test_comm.py::TestCommModbusProtocol::test_split_serial_packet[CommType.SERIAL-socket://localhost:5020] - assert False
FAILED test/sub_transport/test_comm.py::TestCommModbusProtocol::test_serial_poll[CommType.SERIAL-socket://localhost:5020] - AttributeError: 'NoneType' object has no attribute 'connect'
```

Reply:
ahh that's correct, that is still not cleaned up on dev.

But anyhow my comment still stands, deleting variables is not the standard python way (even though it also works).

As I just wrote in another place, we want to have object variables declared in __init__ thus the None, but in reality the way we use the code, it will always. be set....and honestly if somebody takes it and use the class elsewhere it is not our problem.

Reply:
> But anyhow my comment still stands, deleting variables is not the standard python way (even though it also works).

Agreed.  However, I've been unable to find a better way.  🤷🏼‍♂️ 

> As I just wrote in another place, we want to have object variables declared in init thus the None, but in reality the way we use the code, it will always. be set....and honestly if somebody takes it and use the class elsewhere it is not our problem.

There already _is_ a runtime check in `dev`!  This is internal to `pymodbus`, not some external caller
https://github.com/pymodbus-dev/pymodbus/blob/7a3ce95aa73b9e88e4ef5f6d62c83791f3342f1c/pymodbus/transport/transport_serial.py#L41-L42

Go ahead and remove it and watch the tests fail...
```
FAILED test/sub_transport/test_comm.py::TestCommModbusProtocol::test_connected[CommType.SERIAL-socket://localhost:5020] - AttributeError: 'NoneType' object has no attribute 'fileno'
FAILED test/sub_transport/test_comm.py::TestCommModbusProtocol::test_split_serial_packet[CommType.SERIAL-socket://localhost:5020] - assert False
FAILED test/sub_transport/test_comm.py::TestCommModbusProtocol::test_serial_poll[CommType.SERIAL-socket://localhost:5020] - assert False
```

Reply:
I wrote about the theme in general....you know very well transport_serial is imported code that I slowly change to make it fit our needs.

There is a better way, ask mypy to stop complaining or for that sake exclude that file from mypy.

Reply:
> I wrote about the theme in general....

Yes, I understand your point in general. 😃  It's why I'm refactoring the suggestions from @MAKOMO to, for instance, not use `cast()` or `assert()`.  (He's lurking in this thread to see if I have success, I think)

> you know very well transport_serial is imported code that I slowly change to make it fit our needs.

Haha, actually no I don't know that.  😄 You have the entire context for the code in your head, but I only understand bits and pieces.  It's why my contributions are stricter checks that do not rely on understanding the whole codebase.

Another way of thinking about the question...  If code can be explained to `mypy` then it is easier to explain to junior contributors.

Reply:
Yes, I am "lurking" and enjoy the somewhat funny discussion on the use of explicit type specification and automatic validators which seem not to be too important for "internal code" as somehow the developers have some internal (documented?) knowledge that they know (also potential contributors?) and do not violate (as they do not introduce bugs). My experience is different especially with those implicit optionals. They helped me quite a bit to find "null pointer" bugs in my large code base automatically that I would otherwise never have found and it makes me feel comfortable that those checks run on any change I further apply to it. Anyhow, I am following your discussion to learn. Keep going!